[content-service] content store service blob upload/download

corneliusludmann opened the pull request:
First step for issue #2001.

How to test:
- There are integration tests in `test/tests/storage/content-service_test.go`

corneliusludmann commented on components/content-service/pkg/storage/storage.go:
```diff
@@ -289,3 +304,19 @@ func extractTarbal(dest string, src io.Reader) error {
 	return nil
 }
+
+func workspaceObjectName(workspaceName string, name string) string {
+	return fmt.Sprintf("workspaces/%s/%s", workspaceName, name)
+}
```
Seems to be a left-over.

corneliusludmann commented:
@csweichel I started a WIP branch for the content server implementation. Would be great if we could discuss some details tomorrow.

csweichel commented on components/content-service/pkg/storage/storage_test.go:
```diff
@@ -0,0 +1,57 @@
+	}
+	exptectedObjectName := fmt.Sprintf("blobs/%s", name)
+	if objectName != exptectedObjectName {
+		t.Fatalf("unexpected object name: is '%s' but expected '%s'", objectName, exptectedObjectName)
+	}
+}
+
+func TestBlobObjectNameWithSlash(t *testing.T) {
+	name := "my-object-name/with-slash"
+	objectName, err := blobObjectName(name)
+	if err != nil {
+		t.Fatalf("%+v", err)
+	}
+	exptectedObjectName := fmt.Sprintf("blobs/%s", name)
+	if objectName != exptectedObjectName {
+		t.Fatalf("unexpected object name: is '%s' but expected '%s'", objectName, exptectedObjectName)
+	}
+}
+
+func TestBlobObjectNameWithWhitespace(t *testing.T) {
+	name := "object name with whitespace"
+	_, err := blobObjectName(name)
+	if err == nil {
+		t.Fatal("blob name with whitespace should be rejected")
+	}
+	if !strings.Contains(err.Error(), "needs to match regex") {
+		t.Fatalf("%+v", err)
+	}
+}
+
+func TestBlobObjectNameWithInvalidChar(t *testing.T) {
+	name := "object-name-with-invälid-char"
+	_, err := blobObjectName(name)
+	if err == nil {
+		t.Fatal("blob name with invalid char should be rejected")
+	}
+	if !strings.Contains(err.Error(), "needs to match regex") {
+		t.Fatalf("%+v", err)
+	}
+}
```
These tests should be table driven (see the other Go unit tests we have)

corneliusludmann replied:
Changed. Please let me know if you have further hints for improvements here.

csweichel commented on test/tests/storage/content-service_test.go:
```diff
@@ -0,0 +1,197 @@
+	return ok && strings.Contains(st.Message(), "needs to match regex")
+}
+
+func TestUploadUrlReturnsUrl(t *testing.T) {
+	it := integration.NewTest(t)
+	defer it.Done()
+
+	bs := it.API().BlobService()
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	resp, err := bs.UploadUrl(ctx, &api.UploadUrlRequest{OwnerId: gitpodBuiltinUserID, Name: "test-blob"})
+	if err != nil {
+		t.Fatal(err)
+	}
+	url := resp.Url
+	if url == "" {
+		t.Fatal("upload url is empty")
+	}
+	t.Logf("Got URL repsonse: %s", url)
+}
+
+func TestUploadUrlUnknownUser(t *testing.T) {
+	it := integration.NewTest(t)
+	defer it.Done()
+
+	bs := it.API().BlobService()
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	resp, err := bs.UploadUrl(ctx, &api.UploadUrlRequest{OwnerId: "does-not-exist", Name: "test-blob"})
+	if !isErrorNotFound(err) {
+		if err != nil {
+			t.Fatal(err)
+		} else {
+			t.Fatalf("'no found' expected but error is nil, response: %+v", resp)
+		}
+	}
+}
+
+func TestUploadUrlInvalidName(t *testing.T) {
+	it := integration.NewTest(t)
```
Please consider making those tests table driven

corneliusludmann replied:
Changed as well.

corneliusludmann commented on chart/templates/content-service-rolebinding.yaml:
```diff
@@ -0,0 +1,19 @@
+  name: content-service
+roleRef:
+  kind: ClusterRole
+  name: {{ .Release.Namespace }}-ns-psp:restricted-root-user
```
When I apply your suggested change `content-service` does not appear in `kubectl get pods` at all. Since I don't fully understand this change, please let me know whether I should revert this change or not.

corneliusludmann commented:
@csweichel: I've updated this PR:
- Addressed your comments (see above)
- Added an “ensure bucket exists” check in UploadUrl() (that allows uploading blobs when the user bucket has not been created yet.) 7d5b2b3aaed672cd5457887ac19cc02f3682a3ec
- Added a simple quota check 4e7f4ea5e7644dd34dbfca9c5cbeb888c2173178 (free to discuss)
- Rebased on master

csweichel reviewed (approved):
LGTM
